library/core/src/main/java/com/google/android/exoplayer2/source/ConcatenatingMediaSource.java

```diff
@@ -297,6 +304,79 @@ public final synchronized void removeMediaSource(
     }
   }
 
+  /**
+   * Removes a range of {@link MediaSource}s from the playlist, by specifying an initial index
+   * (included) and a final index (excluded).
+   *
+   * <p>Note: when specified range is empty, no actual media source is removed and no exception
+   * is thrown.
+   *
+   * @param fromIndex The initial range index, pointing to the first media source that will be
+   *     removed. This index must be in the range of 0 &lt;= index &lt; {@link #getSize()}.
```

Comment:
Both ranges should be "in the range of 0 &lt;= index &lt;= {@link #getSize()}."  That is, including the getSize index. 
Otherwise you wouldn't be able to do removeMediaSourceRange(getSize(), getSize()) as an edge case.

Reply:
At a given time, "occupied" track indices are 0, 1, ... `getSize() - 1`. `fromIndex` should point to an occupied index, as the initial range index is included in the range. How can one remove the track at index `getSize()`, if there's no track at that index? Just trying to understand

Reply:
As the end index is exclusive, the method must allow getSize() as end index. 

And to support the example you gave in the issue, to keep an element at index N only, it would be nice to be able to call
removeMediaSourceRange(0, N) followed by removeMediaSourceRange(1,getSize()). If this is supposed to remove the last element, the second call will result in removeMediaSourceRange(1, 1) which should be a no-op.

Reply:
We've abandoned the "simplicity" implied by initial issue example when we "gave up" on API tolerance (so to say). Since then, the idea seemed: have a clear API with errors when arguments are not correct. And for `fromIndex` to be equal to `getSize()` seems not correct. Of course we noticed that you cannot call `removeMediaSourceRange(getSize(), getSize())` anymore. And also you cannot call `removeMediaSourceRange(1, getSize())` in issue example, that would throw too.

Staying consistent with allowed param values in signature seems quite a thing. By allowing passing `getIndex()` in `fromIndex` seems to go back to the "tolerant" approach somehow, which is fine anyway. Don't know, your call.

Reply:
I'd say it's best to follow the existing behaviour of [List.subList](https://developer.android.com/reference/java/util/List.html#subList(int,%20int)) and the like. That is, throwing if `startIndex<0`, `endIndex>size` or `startIndex>endIndex`. This ensures the operation is valid, but is also tolerant enough to support common use cases like "remove everything behind element N" with removeRange(N, size) which will work even if N is the last element.

Reply:
Cool. Just to tease a little bit, in case like "remove everything behind element N" when N is the last element, then `N = size - 1`, and `removeMediaSourceRange(size - 1, size)` already works as is. What instead currently throws (and probably looks not tolerant) is `removeMediaSourceRange(size, size)`. Right?

Anyway, changes are coming

Reply:
`removeMediaSourceRange(size - 1, size)` would remove the element at `N = size - 1` though, right? 

And if I'm not mistaken `removeMediaSourceRange(size, size)` won't throw too because List.subList doesn't throw for this case.

Reply:
Q1: correct, current implementation would remove element at `size - 1`, i.e. "the last element"

Q2: `removeMediaSourceRange(size, size)` throws in current (<=> initial version) of this PR. It will not throw after applying your proposed changes, as `List.subList(size, size)` does not throw. I was referring to current implementation.

Cool